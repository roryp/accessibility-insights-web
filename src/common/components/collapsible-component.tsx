// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { css } from '@uifabric/utilities';
import { ActionButton } from 'office-ui-fabric-react';
import * as React from 'react';

export interface CollapsibleComponentProps {
    header: JSX.Element;
    content: JSX.Element;
    contentClassName?: string;
    containerClassName?: string;
}

interface CollapsibleComponentState {
    showContent: boolean;
}

export class CollapsibleComponent extends React.Component<
    CollapsibleComponentProps,
    CollapsibleComponentState
> {
    private readonly iconNameDown = 'ChevronDown';
    private readonly iconNameUp = 'ChevronRight';

    constructor(props: CollapsibleComponentProps) {
        super(props);
        this.state = { showContent: true };
    }

    private onClick = (): void => {
        const newState = !this.state.showContent;
        this.setState({ showContent: newState });
    };

    public render(): JSX.Element {
        const showContent = this.state.showContent;
        let iconName = this.iconNameUp;
        let content: JSX.Element | null = null;

        if (showContent) {
            iconName = this.iconNameDown;
            content = (
                <div className={css(this.props.contentClassName, 'collapsible-content')}>
                    {this.props.content}
                </div>
            );
        }

        return (
            <div className={css(this.props.containerClassName, 'collapsible-component')}>
                <ActionButton
                    className="collapsible"
                    iconProps={{ iconName: iconName, class: 'collapsible-icon' }}
                    onClick={this.onClick}
                    aria-expanded={showContent}
                >
                    <span className="collapsible-title">{this.props.header}</span>
                </ActionButton>
                {content}
            </div>
        );
    }
}
