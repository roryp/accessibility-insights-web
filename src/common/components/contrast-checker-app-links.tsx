// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { windowsProductName } from 'content/strings/application';
import * as React from 'react';
import { NamedFC } from '../react/named-fc';
import { NewTabLink } from './new-tab-link';

export const WindowsContrastCheckerAppLink = NamedFC('WindowsContrastCheckerAppLink', () => (
    <NewTabLink href="https://go.microsoft.com/fwlink/?linkid=2075365">
        {windowsProductName}
    </NewTabLink>
));

export const MacContrastCheckerAppLink = NamedFC('MacContrastCheckerAppLink', () => (
    <NewTabLink href="https://developer.paciellogroup.com/resources/contrastanalyser/">
        Colour Contrast Analyser
    </NewTabLink>
));
