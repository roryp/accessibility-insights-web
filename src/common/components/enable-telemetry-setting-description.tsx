// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';
import { NamedFC } from '../react/named-fc';
import { PrivacyStatementText, PrivacyStatementTextDeps } from './privacy-statement-text';
import { TelemetryNotice, TelemetryNoticeDeps } from './telemetry-notice';

export type EnableTelemetrySettingDescriptionDeps = TelemetryNoticeDeps & PrivacyStatementTextDeps;

export type EnableTelemetrySettingDescriptionProps = {
    deps: EnableTelemetrySettingDescriptionDeps;
};

export const EnableTelemetrySettingDescription = NamedFC<EnableTelemetrySettingDescriptionProps>(
    'EnableTelemetrySettingDescription',
    props => (
        <>
            <TelemetryNotice deps={props.deps} />
            <PrivacyStatementText deps={props.deps} />
        </>
    ),
);
