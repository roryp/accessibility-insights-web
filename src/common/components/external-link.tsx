// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { Link } from 'office-ui-fabric-react';
import * as React from 'react';

import { ActionInitiators } from '../action/action-initiator';
import { NamedFC } from '../react/named-fc';

export type ExternalLinkDeps = {
    actionInitiators: Pick<ActionInitiators, 'openExternalLink'>;
};

export type ExternalLinkProps = {
    deps: ExternalLinkDeps;
    href: string;
    title?: string;
};

export const ExternalLink = NamedFC<ExternalLinkProps>(
    'ExternalLink',
    ({ deps, href, title, children }) => {
        const onClick = e => deps.actionInitiators.openExternalLink(e, { href });
        return (
            <Link
                className="insights-link"
                target="_blank"
                href={href}
                title={title}
                onClick={onClick}
            >
                {children}
            </Link>
        );
    },
);
