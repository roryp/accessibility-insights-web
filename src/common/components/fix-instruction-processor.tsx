// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';
import * as styles from './cards/fix-instruction-color-box.scss';

type ColorMatch = {
    splitIndex: number;
    colorHexValue: string;
};

export class FixInstructionProcessor {
    private readonly colorValueMatcher = `(#[0-9a-f]{6})`;
    private readonly foregroundColorText = 'foreground color: ';
    private readonly foregroundRegExp = new RegExp(
        `${this.foregroundColorText}${this.colorValueMatcher}`,
        'i',
    );
    private readonly backgroundColorText = 'background color: ';
    private readonly backgroundRegExp = new RegExp(
        `${this.backgroundColorText}${this.colorValueMatcher}`,
        'i',
    );

    public process(fixInstruction: string): JSX.Element {
        const matches = this.getColorMatches(fixInstruction);
        return this.splitFixInstruction(fixInstruction, matches);
    }

    private getColorMatches(fixInstruction: string): ColorMatch[] {
        const foregroundMatch = this.getColorMatch(fixInstruction, this.foregroundRegExp);
        const backgroundMatch = this.getColorMatch(fixInstruction, this.backgroundRegExp);

        return [foregroundMatch, backgroundMatch].filter(match => match != null) as ColorMatch[];
    }

    private getColorMatch(fixInstruction: string, colorRegex: RegExp): ColorMatch | null {
        if (!colorRegex.test(fixInstruction)) {
            return null;
        }

        const match = colorRegex.exec(fixInstruction);
        if (match == null || match[1] == null) {
            return null;
        }

        const colorHexValue = match[1];

        return {
            splitIndex: match.index + this.foregroundColorText.length + colorHexValue.length,
            colorHexValue,
        };
    }

    private splitFixInstruction(fixInstruction: string, matches: ColorMatch[]): JSX.Element {
        const sortedMatches = matches.sort((a, b) => a.splitIndex - b.splitIndex);

        if (sortedMatches.length === 0) {
            return <>{fixInstruction}</>;
        }

        let insertionIndex = 0;
        let keyIndex = 0;

        const result: JSX.Element[] = [];

        sortedMatches.forEach(match => {
            const endIndex = match.splitIndex - match.colorHexValue.length;
            const substring = fixInstruction.substring(insertionIndex, endIndex);

            result.push(<span key={`instruction-split-${keyIndex++}`}>{substring}</span>);

            result.push(this.createColorBox(match.colorHexValue, keyIndex++));

            insertionIndex = endIndex;
        });

        const coda = fixInstruction.substr(insertionIndex);

        result.push(<span key={`instruction-split-${keyIndex++}`}>{coda}</span>);

        return (
            <>
                <span aria-hidden="true">{result}</span>
                <span className={styles.screenReaderOnly}>{fixInstruction}</span>
            </>
        );
    }

    private createColorBox(colorHexValue: string, keyIndex: number): JSX.Element {
        return (
            <span
                key={`instruction-split-${keyIndex}`}
                className={styles.fixInstructionColorBox}
                style={{ backgroundColor: colorHexValue }}
            />
        );
    }
}
