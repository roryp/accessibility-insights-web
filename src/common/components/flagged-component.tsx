// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import { FeatureFlagStoreData } from '../types/store-data/feature-flag-store-data';

export interface FlaggedComponentProps {
    featureFlagStoreData: FeatureFlagStoreData;
    featureFlag: string;
    enableJSXElement: JSX.Element;
    disableJSXElement?: JSX.Element;
}

export class FlaggedComponent extends React.Component<FlaggedComponentProps> {
    public render(): JSX.Element | null {
        const flagName = this.props.featureFlag;

        if (this.props.featureFlagStoreData != null && this.props.featureFlagStoreData[flagName]) {
            return this.props.enableJSXElement;
        }

        return this.props.disableJSXElement || null;
    }
}
