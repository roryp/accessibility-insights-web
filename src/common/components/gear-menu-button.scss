// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
@import '../../common/styles/colors.scss';

:global(.ms-Button).gear-menu-button:focus::after {
    border: 1px $neutral-0 solid;
}

:global(.high-contrast-theme) {
    .gear-menu-button {
        color: $neutral-alpha-90;
    }

    .gear-menu-button-callout {
        box-shadow: $always-white 0px 0px 1px 0px !important;
        border-radius: unset;
    }
}
