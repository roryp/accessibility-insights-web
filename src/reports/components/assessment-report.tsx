// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import { ReportModel } from '../assessment-report-model';
import { AssessmentReportBody, AssessmentReportBodyDeps } from './assessment-report-body';
import { AssessmentReportFooter } from './assessment-report-footer';
import { HeaderSection } from './report-sections/header-section';

export type AssessmentReportDeps = AssessmentReportBodyDeps;

export interface AssessmentReportProps {
    deps: AssessmentReportDeps;
    data: ReportModel;
    description: string;
    extensionVersion: string;
    axeVersion: string;
    chromeVersion: string;
}

export class AssessmentReport extends React.Component<AssessmentReportProps> {
    public render(): JSX.Element {
        const targetAppInfo = {
            name: this.props.data.scanDetails.targetPage,
            url: this.props.data.scanDetails.url,
        };

        return (
            <React.Fragment>
                <HeaderSection targetAppInfo={targetAppInfo} />
                <AssessmentReportBody
                    deps={this.props.deps}
                    data={this.props.data}
                    description={this.props.description}
                />

                <AssessmentReportFooter
                    extensionVersion={this.props.extensionVersion}
                    axeVersion={this.props.axeVersion}
                    chromeVersion={this.props.chromeVersion}
                />
            </React.Fragment>
        );
    }
}
