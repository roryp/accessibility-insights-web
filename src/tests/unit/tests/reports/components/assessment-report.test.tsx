// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { shallow } from 'enzyme';
import * as React from 'react';

import { AssessmentReport, AssessmentReportDeps } from 'reports/components/assessment-report';
import { AssessmentReportBuilderTestHelper } from '../../DetailsView/assessment-report-builder-test-helper';

describe('AssessmentReport', () => {
    test('render', () => {
        const deps: AssessmentReportDeps = {
            outcomeTypeSemanticsFromTestStatus: {
                stub: 'outcomeTypeSemanticsFromTestStatus',
            } as any,
        } as AssessmentReportDeps;

        const data = AssessmentReportBuilderTestHelper.getAssessmentReportModel();

        const wrapper = shallow(
            <AssessmentReport
                deps={deps}
                data={data}
                description="test string"
                extensionVersion="ProductVersion"
                axeVersion="axeVersion"
                chromeVersion="chromeVersion"
            />,
        );
        expect(wrapper.getElement()).toMatchSnapshot();
    });
});
