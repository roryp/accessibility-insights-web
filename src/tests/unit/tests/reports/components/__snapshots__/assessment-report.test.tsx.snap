// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssessmentReport render 1`] = `
<React.Fragment>
  <HeaderSection
    targetAppInfo={
      Object {
        "name": "title",
        "url": "url",
      }
    }
  />
  <AssessmentReportBody
    data={
      Object {
        "failedDetailsData": Array [
          Object {
            "displayName": "Assessment1",
            "key": "assessment1",
            "steps": Array [
              Object {
                "defaultMessageComponent": null,
                "header": Object {
                  "description": null,
                  "displayName": "Step2A",
                  "guidanceLinks": Array [],
                  "requirementType": "assisted",
                },
                "instances": Array [
                  Object {
                    "props": Array [
                      Object {
                        "key": "Path",
                        "value": "target1",
                      },
                      Object {
                        "key": "Snippet",
                        "value": "someHtml",
                      },
                    ],
                  },
                ],
                "key": "step2a",
                "showPassingInstances": false,
              },
              Object {
                "defaultMessageComponent": null,
                "header": Object {
                  "description": null,
                  "displayName": "Step4A",
                  "guidanceLinks": Array [],
                  "requirementType": "manual",
                },
                "instances": Array [
                  Object {
                    "props": Array [
                      Object {
                        "key": "Comment",
                        "value": "comment",
                      },
                    ],
                  },
                ],
                "key": "step4a",
                "showPassingInstances": false,
              },
            ],
          },
          Object {
            "displayName": "Assessment2",
            "key": "assessment2",
            "steps": Array [
              Object {
                "defaultMessageComponent": null,
                "header": Object {
                  "description": null,
                  "displayName": "Step2B",
                  "guidanceLinks": Array [],
                  "requirementType": "manual",
                },
                "instances": Array [],
                "key": "step2b",
                "showPassingInstances": true,
              },
            ],
          },
        ],
        "incompleteDetailsData": Array [
          Object {
            "displayName": "Assessment1",
            "key": "assessment1",
            "steps": Array [
              Object {
                "defaultMessageComponent": null,
                "header": Object {
                  "description": null,
                  "displayName": "Step3A",
                  "guidanceLinks": Array [],
                  "requirementType": "assisted",
                },
                "instances": Array [
                  Object {
                    "props": Array [
                      Object {
                        "key": "Path",
                        "value": "target2",
                      },
                      Object {
                        "key": "Snippet",
                        "value": "someHtml",
                      },
                    ],
                  },
                ],
                "key": "step3a",
                "showPassingInstances": false,
              },
            ],
          },
          Object {
            "displayName": "Assessment2",
            "key": "assessment2",
            "steps": Array [
              Object {
                "defaultMessageComponent": null,
                "header": Object {
                  "description": null,
                  "displayName": "Step1B",
                  "guidanceLinks": Array [],
                  "requirementType": "assisted",
                },
                "instances": Array [],
                "key": "step1b",
                "showPassingInstances": true,
              },
            ],
          },
        ],
        "passedDetailsData": Array [
          Object {
            "displayName": "Assessment1",
            "key": "assessment1",
            "steps": Array [
              Object {
                "defaultMessageComponent": null,
                "header": Object {
                  "description": null,
                  "displayName": "Step1A",
                  "guidanceLinks": Array [
                    Object {
                      "href": "link1A",
                      "text": "link1A",
                    },
                  ],
                  "requirementType": "assisted",
                },
                "instances": Array [
                  Object {
                    "props": Array [
                      Object {
                        "key": "Path",
                        "value": "target1",
                      },
                      Object {
                        "key": "Snippet",
                        "value": "someHtml",
                      },
                    ],
                  },
                ],
                "key": "step1a",
                "showPassingInstances": false,
              },
            ],
          },
        ],
        "scanDetails": Object {
          "reportDate": 2000-01-01T00:00:00.000Z,
          "targetPage": "title",
          "url": "url",
        },
        "summary": Object {
          "byPercentage": Object {
            "fail": 50,
            "incomplete": 37,
            "pass": 13,
          },
          "byRequirement": Object {
            "fail": 3,
            "incomplete": 2,
            "pass": 1,
          },
          "reportSummaryDetailsData": Array [
            Object {
              "displayName": "Assessment1",
              "fail": 2,
              "incomplete": 1,
              "pass": 1,
            },
            Object {
              "displayName": "Assessment2",
              "fail": 1,
              "incomplete": 1,
              "pass": 0,
            },
          ],
        },
      }
    }
    deps={
      Object {
        "outcomeTypeSemanticsFromTestStatus": Object {
          "stub": "outcomeTypeSemanticsFromTestStatus",
        },
      }
    }
    description="test string"
  />
  <AssessmentReportFooter
    axeVersion="axeVersion"
    chromeVersion="chromeVersion"
    extensionVersion="ProductVersion"
  />
</React.Fragment>
`;
